BDMS-139: Create `Parameter` table

ksmuczynski opened the pull request:
<!--
- BDMS-139: Create `Parameter` table
-->
### **Why**
_This PR addresses the following problem / context:_
- Need a new `Parameter` table to store centralized metadata about each parameter, such as its category (e.g., 'Metals', 'Field Parameter') and its unique chemical identifier (CAS Number).
- The primary purpose of the `Parameter` table is is to ensure that every observation in the database is linked to a single, authoritative definition of what was measured.
- Need a new `Regulatory_Limit` table to store all the specific, citable standards and limits, regulatory and health-based, associated with a given parameter.
- The `Regulatory_Limit` table allows a single `Parameter` to have multiple different limits set by various agencies (e.g., a federal EPA limit and a state-level NMED limit).

### **How**
_Implementation summary - the following was changed / added / removed:_
- Created a new `Parameter` table and populated with appropriate fields and relationships.
- Created a new `Regulatory_Limit` table and populated with appropriate fields and relationships.
- Updated the `lexicon.json` file to include valid values for the `limit_type` field.
- Added `Parameter` and `RegulatoryLimit` tables to the `__init__.py` file

### **Notes**
_Any special considerations, workarounds, or follow-up work to note?_
- Related validation schema files and test files still need to be created or updated. 
- Transfer script(s) needs to be created.

ksmuczynski committed:
feat: create new `Parameter` model and add it to the `__init__.py` file.

ksmuczynski committed:
feat: create new `regulatory_limit` table.

 refactor: Add valid limit_type values to the `lexicon.json` file.

ksmuczynski committed:
refactor: remove TODO from `regulatory_limit` table.

ksmuczynski committed:
refactor: modify the TODO in the `parameter` table.

ksmuczynski committed:
refactor: modify comment for `limit_source` field.

ksmuczynski committed:
Merge branch 'staging' into parameter_model

ksmuczynski committed:
feat: add `parameter` relationship and foreign key to`Observation` model.

ksmuczynski committed:
Formatting changes

jirhiker commented on db/regulatory_limit.py:
```diff
@@ -0,0 +1,50 @@
+"""
+This table stores the various regulatory or health-based limits for a given
+parameter, sourced from different agencies or standards.
```
@ksmuczynski How is this table going to stay up to date with the authoritative source? Are the regulatory limits shared by the regulators in machine-readable formats? Should we be storing this information, or simply referencing it?

jacob-a-brown replied:
My gut is telling me to fall on the side of referencing it - keeping this table up to date could be difficult and time-consuming. If someone from NMBGMR needs to know this information they can look it up.

ksmuczynski replied:
No, as far as I know, most regulatory standards are shared in human-readable formats, like PDF docs or webpages. Keeping this table up to date would require human responsibility. I think the table itself is still valuable, though, especially for identifying and visualizing exceedances, which I think Sianin mentioned AMP does on occasion. 

If we want to keep this table, I might also add fields for `effective_date` and `end_date` to create a more complete historical record.

jirhiker replied:
I really don't want to have to keep a copy of regulatory limits that must be updated by hand.  I agree about the value of being able to compare to reg limits, but relying on users/manager/admin to maintain is a recipe for disaster

ksmuczynski replied:
I don't know how frequently regs get updated, but I agree it would be better if the table could keep itself up to date. How should we move forward? Should we remove the table for now and introduce it later when we have a better solution for how to maintain it?

jirhiker replied:
I think that we should keep it for now because the benefits outweigh the issues.  we should add regulatory_limits.json to populate with current values. We should reach out to an EPA contact and see if they can provide a machine readable version, or guidance on how to stay up to date

ksmuczynski replied:
@jirhiker If the table is populated from the values in the .json file, what is the benefit of storing values in the .json file versus in the table itself? EPA is only one of many regulatory agencies - should we reach out to all of them?

We don't need to populate with all regulatory agencies and all their standards right now, this table is meant to capture the regs AMP currently checks against and provide the flexibility to add/update in the future.

ksmuczynski replied:
This table is not crucial to the water level functionality and can be populated at a later time.

jirhiker commented on db/parameter.py:
```diff
@@ -0,0 +1,55 @@
+
+    # One-To-Many: A Parameter can have many associated RegulatoryLimits.
+    # If a Parameter is deleted, all its associated limits are deleted as well.
+    reg_limits: Mapped[List["RegulatoryLimit"]] = relationship(
```
change this to `regulatory_limits`

jacob-a-brown commented on db/parameter.py:
```diff
@@ -0,0 +1,55 @@
+
+    __versioned__ = {}
+
+    # --- Columns ---
+    # TODO: Parameter names are currently associated with the 'observed_property' category in the lexicon. Should we update the lexicon category name to 'parameter_name'?
```
That's fine by me. if we do that we should also change the name of the field in the `observation` table.

jacob-a-brown replied:
Would we even need an `observed_property` field? If each observation needs a `parameter_id` then that seems redundant and not necessary.

ksmuczynski replied:
Can different fields reference the same lexicon term? Or does the category have to match the field name? I'm not sure I quite understand how the models use/reference the lexicon, maybe we can discuss early next week.

jacob-a-brown replied:
Different fields can reference the same lexicon term. happy to discuss this more this week

ksmuczynski committed:
refactor: update relationship name from `reg_limit` to `regulatory_limit`

ksmuczynski committed:
Merge remote-tracking branch 'origin/parameter_model' into parameter_model

jacob-a-brown commented on db/parameter.py:
```diff
@@ -0,0 +1,55 @@
+    )
+    cas_number: Mapped[str] = mapped_column(
+        nullable=True,
+        comment="he Chemical Abstracts Service (CAS) registry number, a globally unique identifier for a chemical substance.",
```
"The" is missing a "T" at the beginning

jacob-a-brown commented on db/parameter.py:
```diff
@@ -0,0 +1,55 @@
+        nullable=True,
+        comment="he Chemical Abstracts Service (CAS) registry number, a globally unique identifier for a chemical substance.",
+    )
+    default_unit: Mapped[str] = lexicon_term(
+        nullable=False,
+        comment="The standard, preferred unit for reporting this parameter (e.g., 'ug/L', 'mg/L', 'pH units').",
```
Should there be default units if they may change depending on the matrix? e.g. mg/kg for soil/rock and mg/L for aqueous

jacob-a-brown replied:
or would there be multiple records for the same parameter depending on the matrix?

e.g.

arsenic | water | mg/L
arsenic | soil | mg/kg

would both be different records in this table

ksmuczynski replied:
The purpose of the `Observation.unit` field is to store the actual, raw unit that was reported for a single, specific measurement. The `Parameter.default_unit` field would store the Bureau's preferred standard, or official reporting unit for a particular analyte. The idea behind the `default_unit` field is that it provides us the ability to standardize data. If data are received in a different unit than what the Bureau typically uses, then we can perform necessary unit conversions for plotting/vis/reporting purposes.

Ultimately, the `Parameter.default_unit` field is for ease of unit standardization. Are there other methods or strategies we should consider?

jacob-a-brown replied:
I'm okay to have the `default_unit` field. When we return data we'd probably want to have a unit conversion function so that we can easily and quickly standardize units (and convert values) in the response to the `default_unit`.

What about my above question? If it's the same analyte, but different matrices and units, would there be two different records in the `parameter` table?

ksmuczynski replied:
That's a great question. Yes, there should be two different records in the `Parameter` table, because the `Parameter` table is not just a list of chemical substances, it's a list of analytes, and an analyte takes the matrix into account as well. This is an important distinction as a single chemical, like arsenic, is a completely different analyte from a regulatory and scientific perspective when it's in water versus when it's in soil. They have different units, different analytical methods, and different regulatory limits.

To capture this accurately, I will add a `matrix` column to the `Parameter` table.

ksmuczynski committed:
refactor: fix typo in `cas_number` comment

ksmuczynski committed:
feat: add 'matrix' field and unique constraint to `Parameter` table.

jacob-a-brown commented on db/parameter.py:
```diff
@@ -29,13 +29,17 @@ class Parameter(Base, AutoBaseMixin, ReleaseMixin):
         unique=True,
         comment="The official, full name of the parameter (e.g., 'Arsenic, Dissolved').",
     )
+    matrix: Mapped[str] = lexicon_term(
+        nullable=False,
+        comment="A controlled vocabulary field defining the physical medium the analyte is measured in (e.g., 'Water', 'Soil', 'Air').",
+    )
```
this is also currently recorded in the `sample` table. should we have it in both? or remove it from `sample`?

ksmuczynski replied:
The `matrix` field should be stored in both the `Sample` and the `Parameter` tables. `Sample.matrix` records the factual, physical medium of the sample, while `Parameter.matrix` defines the medium an analyte is intended to be measured in.

ksmuczynski committed:
feat: add `RegulatoryLimit` model to the `db/__init__.py` file so it can be discovered

ksmuczynski committed:
refactor: updated the `lexicon.json` file to use the new normalized structure.

jirhiker committed:
Merge branch 'staging' into parameter_model

jirhiker commented:
@ksmuczynski tests are failing. once they are fixed i will merge PR

ksmuczynski committed:
refactor: fix pytest failures

Create parameter fixtures, update test files, update schema files, update lexicon

jacob-a-brown committed:
fix: don't restrict lexicon strings

jacob-a-brown committed:
fix: eagerly load parameters

jacob-a-brown committed:
fix: use function scopes for fixtures to prevent queue pool limit

jacob-a-brown committed:
fix: coalesce contatinated contact search to prevent null from breaking match

jacob-a-brown committed:
reversion: add pytest back to pre commit hooks

ksmuczynski committed:
refactor: Merge staging

Resolve merge conflicts

jacob-a-brown committed:
refactor: initialize parameters

jacob-a-brown committed:
Merge branch 'parameter_model' of https://github.com/DataIntegrationGroup/NMSampleLocations into parameter_model

jirhiker committed:
Merge branch 'staging' into parameter_model

jirhiker committed:
Update test_observation.py

jirhiker committed:
refactor: update test_observation.py to remove redundant assertions and add value_reason field
